cmd/adapter/query.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/alecthomas/participle"
+	"github.com/alecthomas/participle/lexer"
+	"strconv"
+	"strings"
+)
+
+type (
+	condition struct {
+		Key   string `@("POD"|"CONTAINER"|"FILE") ":"`
+		Value string `@(String|Ident)`
+	}
+
+	expression struct {
+		Or []*orCondition `@@ { "OR" @@ }`
+	}
+
+	orCondition struct {
+		And []*xCondition `@@ { "AND" @@ }`
+	}
+
+	xCondition struct {
+		Not  bool        `(@"NOT")?`
+		Cond *condition  `(@@`
+		Expr *expression `| "(" @@ ")")`
+	}
+
+	query struct {
+		Exp *expression `@@`
+	}
+)
+
+var (
+	qLexer = lexer.Must(lexer.Regexp(`(\s+)` +
+		`|(?P<Keyword>(?i)POD|CONTAINER|FILE|AND|OR|NOT)` +
+		`|(?P<Ident>[a-zA-Z0-9_\.][a-zA-Z0-9_\.\-]*)` + //POSIX "Fully portable filenames"
+		`|(?P<Operator>:|[()])` +
+		`|(?P<String>"([^\\"]|\\.)*"|'[^']*')`,
+	))
+
+	qParser = participle.MustBuild(&query{},
+		participle.Lexer(qLexer),
+		participle.Unquote("String"),
+		participle.CaseInsensitive("Keyword"))
+
+	keyToField = map[string]string{
+		"POD":       "pod",
+		"CONTAINER": "cname",
+		"FILE":      "cid",
+	}
+
+	escaper = strings.NewReplacer("\"", "\\\"")
+)
+
+func parseGrQuery(qs string) (*query, error) {
+	q := &query{}
+	err := qParser.ParseString(qs, q)
+	return q, err
+}
+
+func buildLql(grQuery string, partition string, limit int, offset int) string {
```

Comment:
Same concern - `Lql` does not say much to me. Is it an abbreviation?

Reply:
added more documentation, applied (not yet merged to this branch)